### ISSUE TITLE
[sklearn#33585] ENH Add argument to `SelectFwe` to choose between Bonferroni and Holm method

### ISSUE THREAD

**Issue by godofecht** [flooooooooooow/flow#387]
**Mirrored from [scikit-learn#33585](https://github.com/scikit-learn/scikit-learn/issues/33585)**

Original title: ENH Add argument to `SelectFwe` to choose between Bonferroni and Holm method
Original labels: New Feature, Needs Decision
Created: 2026-03-19T20:56:34Z

---

### Problem

Feature selection is frequently used to reduce the dimensionality of large datasets. To this end, a univariate scoring function can be used to test the hypothesis that a feature correlates with a variable of interest. However, testing a large group of features introduces the multiple comparisons problem where the type I error probability (i.e., the probability of erroneously rejecting a null hypothesis that the feature _does not_ correlate) exceeds the nominal error probability for the whole test family due to the large number of tests performed. A feature may thus appear to correlate with the target variable merely due to noise and end up getting selected erroneously.

`sklearn.feature_selection.SelectFwe` implements the Bonferroni correction which ensures that the actual type I error probability corresponds to the nominal error probability by dividing the significance threshold for each individual p value per feature by the total number of tests. This textbook procedure, however, comes with the risk of increasing the type II error probability, i.e., the error of failing to reject the null hypothesis when it is in fact false. This may result in a failure to detect features although they are informative in reality.

The [Holm method](https://en.wikipedia.org/wiki/Holm%E2%80%93Bonferroni_method) (a.k.a. Holm-Bonferroni method) solves this problem by providing the same protection against type I errors as Bonferroni while reducing the probability of a type II error. In other words, it controls the family-wise error rate as effectively as the Bonferroni correction but is more sensitive to informative features, thereby reducing the risk of missing a feature that in fact does correlate with a target variable. Although the Holm method is more powerful than Bonferroni, it unfortunately is not currently implemented in SelectFwe. 

### Availability in other software packages
The Holm method is commonplace in other (statistics) software like [R](https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/p.adjust) or Python's [`statsmodels` library](https://www.statsmodels.org/dev/generated/statsmodels.stats.multitest.multipletests.html) where it is the default method to control the family-wise error rate.

### Describe your proposed solution

I modified `SelectFwe` so that it now accepts an additional input argument called `fwe_control` of type `str`. The user can choose between "bonf" for Bonferroni, which is its default setting for better backward compatibility, and "holm" for the Holm method. The Holm method sorts all m null hypotheses of a test family by their corresponding p values in ascending order $p_1<p_2<...<p_m$ and applies a modified Bonferroni correction in a step-wise fashion. It rejects the first null hypothesis if $p_1<\frac{\alpha}{m}$. If this null hypothesis is rejected, it rejects the second null hypothesis if $p_2<\frac{\alpha}{m−1}$. (The intuition is that the first 

...(truncated)

---

## Flow context

This issue is mirrored from scikit-learn to explore whether Flow encounters the same problem or can approach it differently. Add analysis below.

**Comment by godofecht** [comment]
Moved to [8](https://github.com/godofecht/flow-scikit/issues/8) in the flow-scikit repository. Further library-specific triage and implementation will happen there.
